### DISCUSSION TITLE
jquery problems with thuliteio / doks / hugo and the dearflip viewer (or any other jquery stuff...

### DISCUSSION THREAD

**Discussion by cnichte** [thuliteio/doks#1287] [Q&A] [answered]
Hi :-)

I have to start by saying that a simple basic example outside of ` thuliteio / doks / hugo` works perfectly as you can see here:

```html
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>dFlip Test</title>

<!-- Flipbook StyleSheets -->
<link href="https://dflip.test/css/dflip.min.css" rel="stylesheet" type="text/css">

</head>
<body>
<div class="_df_book" id="flipbok_example" source="filename.pdf"></div>

<!-- Scripts -->
<script src="https://dflip.test/js/libs/jquery-3.7.1.js" type="text/javascript"></script>
<script src="https://dflip.test/js/dflip.min.js" type="text/javascript"></script>

</body>
</html>
```

That gives me this nice view:

![Bildschirmfoto von Google Chrome am 15_09_2024, 16_08_59](https://github.com/user-attachments/assets/30c7fe90-a6d7-4584-865b-8766deabf34d)

This describes what I want: The PDF viewer should be displayed.  What works fine outside ` thuliteio / doks / Hugo `, is a problem with ` thuliteio / doks / Hugo `:  

```
 jQuery is not found
```

## What I have done

In `layouts/partials/head/resource-hints.html`:

```
{{ $styles := resources.Get "dflip/css/dflip.css" | minify }}
<link rel="stylesheet" href="{{ $styles.Permalink }}" media="screen">
```

in `layouts/partials/footer/script-footer-custom.html`:

```
{{ partial "footer/esbuild" (dict "src" "js/jquery-3.7.1.js" "transpile" false) -}}
{{ partial "footer/esbuild" (dict "src" "dflip/js/dflip.js" "transpile" false) -}}
```

All required resources are in place in the `assets` folder.

Then I have  a page with a shortcode that inserts the display component:

```html
 <div class="_df_book" id="flipbok_example" source="Test.pdf"></div>
```

The created output ist - stripped down to the relevant parts for a better overview:

```html
<!doctype html>
<html>
<head>
    <link rel="stylesheet" href="http://localhost:1313/dflip/css/dflip.min.css" media="screen">
</head>
<body>

     <div class="_df_book" id="flipbok_example" source="Test.pdf"></div>

    <script  src="/js/jquery-3.7.1.e5dd773c71859e5734abe80e7ae68787f5f5bfe909dce408ee68d5ce4135f1e4.js"
  integrity="sha256-5d13PHGFnlc0q&#43;gOeuaHh/X1v&#43;kJ3OQI7mjVzkE18eQ="></script>

<script  src="/dflip/js/dflip.fa58fbe4d511a32899090f96ce3b1234ae5aae069b288d802d9107bb3b7cd5b4.js"  integrity="sha256-&#43;lj75NURoyiZCQ&#43;WzjsSNK5argabKI2ALZEHuzt81bQ="></script>
</body>
</html>
```

What I get is an `<stdin>:441 Uncaught ReferenceError: jQuery is not defined`:

![Bildschirmfoto von Google Chrome am 15_09_2024, 16_05_11](https://github.com/user-attachments/assets/b64fe0f8-eeaa-47ce-aba1-c3119c0d9228)

Do you have any idea what could be causing this? Any advice is welcome and helpful.

Best regards - Carsten

**Comment by h-enk** [comment]
Looks like `/dflip/js/dflip.fa58fbe4d511a32899090f96ce3b1234ae5aae069b288d802d9107bb3b7cd5b4.js` finished loading before `/js/jquery-3.7.1.e5dd773c71859e5734abe80e7ae68787f5f5bfe909dce408ee68d5ce4135f1e4.js"
  integrity="sha256-5d13PHGFnlc0q&#43;gOeuaHh/X1v&#43;kJ3OQI7mjVzkE18eQ=`

Try with the `defer attribute`?

```html
{{ partial "footer/esbuild" (dict "src" "js/jquery-3.7.1.js" "load" "defer" "transpile" false) -}}
{{ partial "footer/esbuild" (dict "src" "dflip/js/dflip.js" "load" "defer" "transpile" false) -}}
```

See also: [`<script>`: The Script element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script)

**Comment by cnichte** [comment]
Thanks for the quick feedback.  No, `defer` has unfortunately the same effect:

![Bildschirmfoto von Google Chrome am 16_09_2024, 12_48_05](https://github.com/user-attachments/assets/f3d72c92-dced-4ded-9bf1-edbd6d348f79)

I noticed in my minimalistic working example, that additional scripts seem to be injected:

The source code looks like this:

```html

<script src="https://dflip.test/js/libs/jquery-3.7.1.js" type="text/javascript"></script>
<script src="https://dflip.test/js/dflip.js" type="text/javascript"></script>
</body>
``` 

But in the browser it looks like this. Miraculously, two scripts have been added:

![Bildschirmfoto von Google Chrome am 16_09_2024, 12_43_43](https://github.com/user-attachments/assets/200365b0-ef84-4e87-90b5-b7abdfba8a84)

Could it be that in the `thuliteio / doks / Hugo` manner the  browser is preventing this? I would try to add the two lines manually. However, the scripts are minified. Can I tell Hayas to just pass the files / lines through as they are without minifying them? ....

```html
<script async data-cfasync="false" src="https://dflip.test/is/Libs/pdf.min.js?ver=2.2.548pdfver=default>/script>
<script async data-cfasync="false" src-"https:///dflip.test/is//Ubs/three.min-js/ver=2.2.54"></script>
```

Thanks for your patience:-D

**Reply by h-enk** [reply]
Another thing to try: load jQuery in the head (this is also how DearFlip/WordPress does it).

So, add to `layouts/partials/head/script-header.html` (first try with `"load" "async" — if you still experience loading issues, use "defer"`):

```html
{{ partial "footer/esbuild" (dict "src" "js/jquery-3.7.1.js" "load" "async" "transpile" false) -}}
```

And in `layouts/partials/footer/script-footer-custom.html`:

```html
{{ partial "footer/esbuild" (dict "src" "dflip/js/dflip.js" "load" "defer" "transpile" false) -}}
```

**Reply by h-enk** [reply]
I think in the `head` should do it — https://www.w3schools.com/jquery/jquery_get_started.asp

**Reply by cnichte** [reply]
Thank you, 

I followed this example: https://jquery.dearflip.com/docs/file-structure-template/ and, as I said, it works without any problems outside thuliteio / hyas / Hugo universe.

Following your suggestion, i also tried jQuery in the header with async and then with defer, and dearFlip in the footer.  Always the same error. I have now put the test online to have a look: https://carsten-nichte.de/info/flipbook/

The scripts are all in place. I just lack the imagination and the knowledge of what is going wrong there :-)

**Comment by h-enk** [comment, marked as answer]
This works (tested)

Add to `layouts/partials/footer/script-footer-custom.html`:

```html
{{ $jquery := resources.Get "js/jquery.min.js" }}
<script src="{{ $jquery.RelPermalink }}"></script>

{{ partial "footer/esbuild" (dict "src" "js/dflip.js" "load" "async" "transpile" false) -}}
```

**Reply by cnichte** [reply]
That is basically the solution.

I added the following to `layouts/partials/footer/script-footer-custom.html`:

```html
{{/* I have to put dflip in the assets folder and load booth in this way: */ -}}
{{ $jquery := resources.Get "dflip/js/libs/jquery.min.js" }}
<script src="{{ $jquery.RelPermalink }}"></script>

{{ $dflip := resources.Get "dflip/js/dflip.min.js" }}
<script async data-cfasync="false" src="{{ $dflip.RelPermalink }}"></script>

{{/* ----------------------------------------------------------
The following two scripts are dynamically injected from dflip...
 so i have to provide these via `static` folder.
<script async="" data-cfasync="false" src="http://localhost:1313/dflip/js/libs/pdf.min.js?ver=2.2.54&amp;pdfver=default"></script>
<script async="" data-cfasync="false" src="http://localhost:1313/dflip/js/libs/three.min.js?ver=2.2.54"></script> 
*/ -}}
```

After that the whole thing works like a charm. Thanks!

**Reply by h-enk** [reply]
Great you got it working!
